WebDriver Bidi: add initial implementations for `script` and `browsingContext` commands.

burg opened the pull request:
#### fe32a3d1f71f9c7594351773210691982a496514
<pre>
WebDriver Bidi: add initial implementations for `script` and `browsingContext` commands.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281943">https://bugs.webkit.org/show_bug.cgi?id=281943</a>
&lt;<a href="https://rdar.apple.com/problem/138880813">rdar://problem/138880813</a>&gt;

Reviewed by Devin Rousso.

In order to unblock more testing, implement simplified versions of core commands that are
used by most tests. This includes script evaluation, navigation, activate, create, close, and getTree.
We use the strategy of calling existing WebAutomationSession methods where possible and bridging
the results as necessary. Add FIXMEs for known missing pieces of argument and result handling.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
Add BidiScript.json.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
Give the Bidi processor access to protected session methods like `handleForWebPageProxy`.
We might figure out a better way to do this in the future once frame handle support is added.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
(WebKit::WebDriverBidiProcessor::activate): Added.
(WebKit::WebDriverBidiProcessor::close): Added.
(WebKit::browsingContextPresentationFromCreateType): Added.
(WebKit::WebDriverBidiProcessor::create): Added.
(WebKit::WebDriverBidiProcessor::getTree): Added.
Instead of calling `page-&gt;getWindowFrameWithCallback()` repeatedly, we can just iterate pages and frames.

(WebKit::pageLoadStrategyFromReadinessState): Added.
(WebKit::WebDriverBidiProcessor::navigate): Added.
(WebKit::WebDriverBidiProcessor::reload): Added.
(WebKit::WebDriverBidiProcessor::callFunction): Added.
(WebKit::WebDriverBidiProcessor::evaluate): Added.

* Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiScript.json: Added.
Add links to specification and tests where applicable. Add better descriptions. Add new types and commands.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Drive-by, add BiDi domain input files and generated combined file to the Xcode project.
This makes it easier to refer back to the input file while working in Xcode.

* Source/WebKit/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
Add baseline expectation for existing SaferCPP failure related to generated protocol code. This matches
the expectations for AutomationProtocolObjects.h. Improving this is to be addressed as a separate fix.

Canonical link: <a href="https://commits.webkit.org/293479@main">https://commits.webkit.org/293479@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b954ff4f29787920827832661ae2602172723d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99022 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18660 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8906 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104151 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49615 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18954 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27109 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/75381 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32509 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102026 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/14415 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89437 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55742 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14207 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48991 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84140 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7487 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106517 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26119 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19047 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/84347 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26518 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85635 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83850 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21266 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/28524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6187 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19856 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26091 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31258 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25892 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29216 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27466 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lauromoura commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,19 +117,177 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
         ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
 
-    ASYNC_FAIL_WITH_PREDEFINED_ERROR(NotImplemented);
+    // FIXME: implement `frameHandle` to activate non-top level browsing contexts.
```
Per the spec, we should return `"invalid argument"` when trying to activate non-top level browsing contexts.

Step 3 in https://w3c.github.io/webdriver-bidi/#command-browsingContext-activate : 
> If navigable is not a [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable), return [error](https://w3c.github.io/webdriver/#errors) with [error code](https://w3c.github.io/webdriver/#dfn-error-code) [invalid argument](https://w3c.github.io/webdriver/#dfn-invalid-argument).
```suggestion
    // FIXME: detect non-top level browsing contexts, returning `invalid argument`
```

lauromoura commented on Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
```diff
@@ -0,0 +1,174 @@
+            "enum": [ "success", "exception" ]
+        },
+        {
+            "id": "EvaluateResultOwnership",
```
Should we use `ResultOwnership` instead, to match the spec term? It will also be used serializing data that comes from the browser through the `script.Channel` type.

https://w3c.github.io/webdriver-bidi/#type-script-ResultOwnership

```suggestion
            "id": "ResultOwnership",
```

lauromoura commented on Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
```diff
@@ -0,0 +1,174 @@
+                "arraybuffer",
+                "date",
+                "error",
+                "function",
```
Nit: missing `"generator"`?
https://w3c.github.io/webdriver-bidi/#type-script-RemoteValue

```suggestion
                "function",
                "generator",
```

lauromoura commented on Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
```diff
@@ -0,0 +1,174 @@
+                "nodelist",
+                "object",
+                "promise",
+                "proxy",
```
Nit: missing `"regexp"`?
```suggestion
                "proxy",
                "regexp",
```

lauromoura commented on Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
```diff
@@ -0,0 +1,174 @@
+            "properties": [
+                { "name": "realm", "$ref": "BidiScript.Realm", "optional": true },
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext", "optional": true },
+                { "name": "value", "type": "any" }
```
What's this `"value"` key? Shouldn't it be the optional `sandbox` for `script.ContextTarget` specialization?

https://w3c.github.io/webdriver-bidi/#type-script-Target

```suggestion
                { "name": "sandbox", "type": "string", "optional": true }
```

burg replied:
Uh, no idea! Probably added to wrong command.

lauromoura commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -22,17 +42,74 @@
+                { "name": "userContext", "$ref": "BidiBrowser.UserContext", "optional": true }
+            ],
+            "returns": [
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext", "description": "The identifier of the browsing context to navigate." }
```
```suggestion
                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext", "description": "The identifier of the newly created browsing context." }
```

lauromoura commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,19 +117,177 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    // FIXME: implement `promptUnload` option.
```
```suggestion
    // FIXME: implement `promptUnload` option.
    // FIXME: raise `invalid argument` if `browsingContext` is not a top-level traversable
```

lauromoura reviewed (commented):
Looking good in general, but some questions on some names in comparison to the spec.

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
```diff
@@ -279,9 +284,11 @@ class WebAutomationSession final : public API::ObjectImpl<API::Object::Type::Aut
     void didDestroyFrame(WebCore::FrameIdentifier);
 
-private:
+protected:
```
is this actually necessary since you've declared `friend class WebDriverBidiProcessor;`?

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,19 +117,183 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
-    ASYNC_FAIL_WITH_PREDEFINED_ERROR(NotImplemented);
+    RefPtr webPageProxy = session->webPageProxyForHandle(browsingContext);
+    if (!webPageProxy)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(WindowNotFound);
```
i wonder if it'd be helpful to have a `ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF` so that this can just be written as
```cpp
ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!webPageProxy, WindowNotFound);
```
personally i think that'd be cleaner

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,19 +117,183 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+    if (requestedCreateType != createType)
+        createType = Inspector::Protocol::BidiBrowsingContext::CreateType::Window;
+
+    auto presentationHint = createType == Inspector::Protocol::BidiBrowsingContext::CreateType::Tab ? Inspector::Protocol::Automation::BrowsingContextPresentation::Tab : Inspector::Protocol::Automation::BrowsingContextPresentation::Window;
```
this seems unnecessarily complex
```cpp
auto presentationHint = requestedCreateType == Inspector::Protocol::BidiBrowsingContext::CreateType::Tab ? Inspector::Protocol::Automation::BrowsingContextPresentation::Tab : Inspector::Protocol::Automation::BrowsingContextPresentation::Window;
```
or more ideally we'd have a `static` method to convert these values

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,19 +117,183 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+        }
+
+        auto [resultContext, resultPresentation] = WTFMove(result.value());
+        callback(resultContext);
```
should this be a `WTFMove(resultContext)` since it's moved out of `result.value()`?

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,19 +117,183 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+        }
+    }
+
+    Ref<JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>> infos = JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create();
```
why not do this all in one `for`?  is there some reason we have to have the intermediate `Vector`?  or can we just add to this JSON directly?

burg replied:
The page iteration code is just a placeholder, we will want to share this code in several places and also iterate over frames. We might want to ask the client (Safari etc) to filter out prewarmed processes, extension pages, about:blank style pages, etc.

 I want to separate page and frame tree iteration and filtering, from the mechanical building up of the protocol objects. I'll make this more clear with builders/functors once the opportunity for code use arises.

dcrousso replied:
it still seems like unless we plan to change the order or something like that it would be easy to combine it all into a single `for`
```cpp
for (Ref process : session->protectedProcessPool()->processes()) {
    for (Ref page : process->pages()) {
        if (!page->isControlledByAutomation())
            continue;
        if (/* some other check */)
            continue;
        infos->addItem(...);
    }
}
```
but i suppose what you have is also fine

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,19 +117,183 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    PageLoadStrategy strategy = PageLoadStrategy::None;
+    switch (optionalReadinessState.value_or(Inspector::Protocol::BidiBrowsingContext::ReadinessState::None)) {
```
can we make this into a `static` function so that we can avoid writing it again below?

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -136,11 +306,84 @@ Inspector::Protocol::ErrorStringOr<void> WebDriverBidiProcessor::close()
+    if (arguments) {
+        for (auto argument : *arguments)
+            outArguments->pushValue(argument.copyRef());
+    }
```
why do we need to create a copy of `arguments`?

burg replied:
`RefPtr<JSON::Object>` is an optional argument, but the call to `WebAutomationSession::evaluateJavascriptFunction` wants a non-null, possibly empty, `JSON::Object`.

dcrousso replied:
couldn't we just do something like this?
```cpp
arguments ? arguments.releaseNonNull() : JSON::Object::create()
```

dcrousso commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
```diff
@@ -22,17 +42,74 @@
+            "parameters": [
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext", "description": "The identifier of the browsing context to activate.  It is an error to pass a non-top level traversable." }
+            ],
+        "async": true
```
all of these should be indented

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,19 +117,183 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    // FIXME: implement `promptUnload` option.
```
if we're not using `optionalPromptUnload` right now, is there a reason we're handling it?  id much prefer not having the parameter at all (both here and in the protocol `.json`) if it's not supported

ditto with many other FIXME below (and to be clear i have no issue with having a FIXME comment, just that we also shouldn't support the code if it's not implemented)

burg replied:
Unlike inspector protocol and automation protocol, WebDriver Bidi has a spec-defined list of arguments. So I prefer the approach taken. Based on the review feedback in this PR, I am eager to add 'spec' and 'bug' fields to the .json domain files.

Currently, I don't have a sense of how fine-grained we want to be with bugzilla links for FIXMEs. I will update the .json with spec links and bug links separately.

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,36 +117,262 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+        }
+    }
+
+    Ref<JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>> infos = JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create();
```
Style: `auto infos =`

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,36 +117,262 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
 
-    ASYNC_FAIL_WITH_PREDEFINED_ERROR(NotImplemented);
+    RefPtr webPageProxy = session->webPageProxyForHandle(browsingContext);
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!webPageProxy, WindowNotFound);
```
omg this is looks **_SO_** much better! 🤩

dcrousso commented on Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
```diff
@@ -111,36 +117,262 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+
+// https://www.w3.org/TR/webdriver/#dfn-session-page-load-timeout
+static const Seconds defaultPageLoadTimeout = 300_s;
+static const ReadinessState defaultReadinessState = ReadinessState::None;
```
i think these should be `constexpr`

dcrousso commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
```diff
@@ -1,12 +1,29 @@
     "condition": "ENABLE(WEBDRIVER_BIDI)",
+    "description": "The browser module contains commands for managing the remote end browser process.",
+    "spec": "https://w3c.github.io/webdriver-bidi/#module-browser",
+    "tests": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/browser",
```
NIT: `"tests"` makes me think it's a link to something else in WebKit source, so how about `"wpt"` or `"testURL"` or something to clarify that it's external?

dcrousso reviewed (approved):
im really not a fan of adding arguments to the protocol that we don't support, but i realize that this probably wont ship until full/majority support is implemented, so im willing to acquiesce for now

burg committed:
WebDriver Bidi: add initial implementations for `script` and `browsingContext` commands.
https://bugs.webkit.org/show_bug.cgi?id=281943
<rdar://problem/138880813>

Reviewed by Devin Rousso.

In order to unblock more testing, implement simplified versions of core commands that are
used by most tests. This includes script evaluation, navigation, activate, create, close, and getTree.
We use the strategy of calling existing WebAutomationSession methods where possible and bridging
the results as necessary. Add FIXMEs for known missing pieces of argument and result handling.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
Add BidiScript.json.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
Give the Bidi processor access to protected session methods like `handleForWebPageProxy`.
We might figure out a better way to do this in the future once frame handle support is added.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
(WebKit::WebDriverBidiProcessor::activate): Added.
(WebKit::WebDriverBidiProcessor::close): Added.
(WebKit::browsingContextPresentationFromCreateType): Added.
(WebKit::WebDriverBidiProcessor::create): Added.
(WebKit::WebDriverBidiProcessor::getTree): Added.
Instead of calling `page->getWindowFrameWithCallback()` repeatedly, we can just iterate pages and frames.

(WebKit::pageLoadStrategyFromReadinessState): Added.
(WebKit::WebDriverBidiProcessor::navigate): Added.
(WebKit::WebDriverBidiProcessor::reload): Added.
(WebKit::WebDriverBidiProcessor::callFunction): Added.
(WebKit::WebDriverBidiProcessor::evaluate): Added.

* Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiScript.json: Added.
Add links to specification and tests where applicable. Add better descriptions. Add new types and commands.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Drive-by, add BiDi domain input files and generated combined file to the Xcode project.
This makes it easier to refer back to the input file while working in Xcode.

* Source/WebKit/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
Add baseline expectation for existing SaferCPP failure related to generated protocol code. This matches
the expectations for AutomationProtocolObjects.h. Improving this is to be addressed as a separate fix.

Canonical link: https://commits.webkit.org/293479@main
